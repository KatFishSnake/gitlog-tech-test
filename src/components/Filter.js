import { useState } from "react";
// import { debounce } from "lodash";

const Filter = ({ onQuery }) => {
  const [searchQuery, setSearchQuery] = useState("");

  /**
   * Would be this if proposed changes accepted
   */
  // const onSearchDebounced = useCallback(
  //   debounce((e) => {
  //     onQuery(e.target.value);
  //   }, 500),
  //   [onQuery]
  // );

  // const handleSearch = (event) => {
  //   event.persist();
  //   setSearchQuery(event.target.value);
  //   onSearchDebounced(event);
  // };

  const handleSearch = () => {
    onQuery(searchQuery);
  };

  return (
    <div className="w-full h-16 p-4 text-md bg-grey-200 z-1 shadow-md z-1">
      <input
        type="search"
        name="commitquery"
        className="py-1 text-sm focus:outline-none border border-grey-300"
        autoComplete="off"
        placeholder="Filter by commit messageâ€¦"
        value={searchQuery}
        onChange={(e) => {
          setSearchQuery(e.target.value);
        }}
      />
      <button onClick={handleSearch}>Filter</button>
    </div>
  );
};

export default Filter;
