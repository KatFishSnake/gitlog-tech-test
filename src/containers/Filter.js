import { useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import { sizes } from "../style-vars";

const SearchField = styled.input`
  width: 100%;

  @media ${sizes.md} {
    max-width: 300px;
  }
`;

const SearchTriggerButton = styled.button`
  background: linear-gradient(180deg, #faf2f2 0%, #ede5e5 100%);
  border: 1px solid #d0cccc;
`;

const Filter = ({ onQuery }) => {
  const [searchQuery, setSearchQuery] = useState("");

  /**
   * Would be this if proposed changes accepted
   */
  // const onSearchDebounced = useCallback(
  //   debounce((e) => {
  //     onQuery(e.target.value);
  //   }, 500),
  //   [onQuery]
  // );

  const handleSearch = (event) => {
    setSearchQuery(event.target.value);
    // On clear bypass the confirm flow
    if (event.target.value.length === 0) onQuery("");
  };

  const handleFilterConfirm = () => {
    onQuery(searchQuery);
  };

  return (
    <div className="w-full h-16 px-4 text-md bg-l-gray flex items-center flex-none">
      <SearchField
        type="search"
        name="commitquery"
        className="py-1 text-xs focus:outline-none border border-gray-300 pl-3 pr-1 mr-2 h-7 align-top"
        autoComplete="off"
        placeholder="Filter by commit messageâ€¦"
        value={searchQuery}
        onChange={(e) => {
          handleSearch(e);
        }}
      />
      <SearchTriggerButton
        onClick={handleFilterConfirm}
        className="px-6 text-xs py-1 h-7 align-top"
      >
        Filter
      </SearchTriggerButton>
    </div>
  );
};

Filter.propTypes = {
  onQuery: PropTypes.func,
};

export default Filter;
