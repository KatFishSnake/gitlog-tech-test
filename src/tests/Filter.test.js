import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Filter from "../containers/Filter";

test("Renders Filter consistently", () => {
  const { container } = render(<Filter />);
  expect(container).toMatchSnapshot();
});

test("Correctly passes value to parent on query", () => {
  let expectedValue = "This is my query";
  let query = null;
  const { container, getByPlaceholderText, getByText } = render(
    <Filter
      onQuery={(val) => {
        query = val;
      }}
    />
  );

  const input = getByPlaceholderText("Filter by commit message…");

  userEvent.type(input, expectedValue);
  userEvent.click(getByText("Filter"));

  expect(query).toBe(expectedValue);
});

test("Correctly passes value to parent on clear, without triggering Filter button", () => {
  let expectedValue = "This is my query";
  let query = null;
  const { container, getByPlaceholderText } = render(
    <Filter
      onQuery={(val) => {
        query = val;
      }}
    />
  );

  const input = getByPlaceholderText("Filter by commit message…");

  userEvent.type(input, expectedValue);
  userEvent.clear(input);

  expect(query).toBe("");
});
