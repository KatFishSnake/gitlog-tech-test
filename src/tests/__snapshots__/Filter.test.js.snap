// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Filter consistently 1`] = `
<div>
  <div
    class="w-full h-16 px-4 text-md bg-l-gray flex items-center flex-none"
  >
    <input
      autocomplete="off"
      class="sc-bdfBwQ hluRTZ py-1 text-xs focus:outline-none border border-gray-300 pl-3 pr-1 mr-2 h-7 align-top"
      name="commitquery"
      placeholder="Filter by commit messageâ€¦"
      type="search"
      value=""
    />
    <button
      class="sc-gsTCUz kAPWUU px-6 text-xs py-1 h-7 align-top"
    >
      Filter
    </button>
  </div>
</div>
`;
